core/tx_list.go

```diff
@@ -0,0 +1,331 @@
+			*l.index = append(*l.index, nonce)
+		}
+		heap.Init(l.index)
+	}
+	l.cache = nil
+
+	return removed, invalids
+}
+
+// Cap places a hard limit on the number of items, returning all transactions
+// exceeding tht limit.
+func (l *txList) Cap(threshold int) types.Transactions {
+	// Short circuit if the number of items is under the limit
+	if len(l.items) < threshold {
+		return nil
+	}
+	// Otherwise gather and drop the highest nonce'd transactions
+	var drops types.Transactions
+
+	if l.strict {
+		// In strict mode, just gather top down from last to first
+		for len(l.items) > threshold {
+			if tx, ok := l.items[l.last]; ok {
+				drops = append(drops, tx)
+				delete(l.items, l.last)
+				l.last--
+			}
+		}
+	} else {
+		// In gapped mode it's expensive: we need to sort and drop like that
+		sort.Sort(*l.index)
+		for size := len(l.items); size > threshold; size-- {
+			drops = append(drops, l.items[(*l.index)[size-1]])
+			delete(l.items, (*l.index)[size-1])
+			*l.index = (*l.index)[:size-1]
+		}
+		heap.Init(l.index)
+	}
+	l.cache = nil
+
+	return drops
+}
+
+// Remove deletes a transaction from the maintained list, returning whether the
+// transaction was found, and also returning any transaction invalidated due to
+// the deletion (strict mode only).
+func (l *txList) Remove(tx *types.Transaction) (bool, types.Transactions) {
+	nonce := tx.Nonce()
+	if _, ok := l.items[nonce]; ok {
+		// Remove the item and invalidate the sorted cache
+		delete(l.items, nonce)
+		l.cache = nil
+
+		// Remove all invalidated transactions (strict mode only!)
+		invalids := make(types.Transactions, 0, l.last-nonce)
+		if l.strict {
+			for i := nonce + 1; i <= l.last; i++ {
+				invalids = append(invalids, l.items[i])
+				delete(l.items, i)
+			}
+			l.last = nonce - 1
+		} else {
+			// In gapped mode, remove the nonce from the index but honour the heap
+			for i := 0; i < l.index.Len(); i++ {
+				if (*l.index)[i] == nonce {
+					heap.Remove(l.index, i)
+					break
+				}
+			}
+		}
+		// Figure out the new highest nonce
+		return true, invalids
+	}
+	return false, nil
+}
+
+// Ready retrieves a sequentially increasing list of transactions starting at the
+// provided nonce that is ready for processing. The returned transactions will be
+// removed from the list.
+//
+// Note, all transactions with nonces lower that start will also be returned to
+// prevent getting into and invalid state. This is not something that should ever
+// happen but better to be self correcting than failing!
+func (l *txList) Ready(start uint64) types.Transactions {
+	var txs types.Transactions
+	if l.strict {
+		// In strict mode make sure we have valid transaction, return all contiguous
+		if l.first > start {
+			return nil
+		}
+		for {
+			if tx, ok := l.items[l.first]; ok {
+				txs = append(txs, tx)
+				delete(l.items, l.first)
+				l.first++
+				continue
+			}
+			break
+		}
+	} else {
+		// In gapped mode, check the heap start and return all contiguous
```

Comment:
In the case of executable pending transactions, we know they are contiguous, so representing the order via a heap would be very suboptimal imho. It would bump the complexity of all the operations from O(1) to O(log n).